### PR TITLE
DOC: Add cookbook entry using callable method for DataFrame.corr

### PR DESCRIPTION
**Pull request by dsaxton**
Provides a cookbook entry using the callable method option for `DataFrame.corr` (PR #22684) to calculate a distance correlation matrix.  (Related: issue #22402)

### CODE REVIEW

File: doc/source/cookbook.rst [hunk 1]
```diff
@@ -1225,6 +1225,42 @@ Computation
...
+
+.. ipython:: python
+
+    def distcorr(x, y):
+        n = len(x)
+        a = np.zeros(shape=(n, n))
+        b = np.zeros(shape=(n, n))
+
+        for i in range(n):
+            for j in range(i+1, n):
```
Comment (datapythonista) [inline doc/source/cookbook.rst]: missing spaces around `+`

File: doc/source/cookbook.rst [hunk 2]
```diff
@@ -1225,6 +1225,42 @@ Computation
...
+                a[i, j] = abs(x[i] - x[j])
+                b[i, j] = abs(y[i] - y[j])
+
+        a = a + a.T
+        b = b + b.T
```
Comment (datapythonista) [inline doc/source/cookbook.rst]: I think `a += a.T` is more Pythonic

**Review by datapythonista — CHANGES_REQUESTED** [review]
Is the distance correlation you provided different from: https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.spatial.distance.correlation.html ?

While the example looks interesting, it doesn't seem enough related to pandas to be worth adding it to our cookbook.

If the function in scipy is the same as the one you implemented, it could be nice having this example (using the scipy function) in the `DataFrame.corr` docstring.

What do you think?

**Comment by dsaxton** [comment]
The function from `scipy.spatial.distance` is taking one minus the ordinary Pearson correlation, so this is a different calculation.  The cookbook entry was requested by @TomAugspurger in #22402 and I think the goal was just to show the flexibility of allowing callable methods in `DataFrame.corr` while also being somewhat interesting in itself.

**Comment by TomAugspurger** [comment]
I think this will be good to have.

We'll need to wait for https://github.com/pandas-dev/pandas/pull/22684 first.

**Comment by datapythonista** [comment]
Sorry @dsaxton, I didn't see the related issue (next time it may be a good idea to leave the template we have in the PR description, it makes it easier to see there is a related issue).

**Comment by gfyoung** [comment]
@dsaxton : Can you rebase onto `master` to fix the build issues?

**Comment by datapythonista** [comment]
@dsaxton seems like something went wrong with git, and your PR contains many unrelated changes. Can you take a look?

In general that could be fixed by performing being in the PR branch:
```
git fetch upstream
git merge upstream/master
git reset --soft upstream/master
git commit -m "PR description"
git push -f
```

Thanks!

**Comment by dsaxton** [comment]
@datapythonista Sorry about that, could this has been caused by my (attempted) rebase?  I'll try the commands you're suggesting.

**Comment by datapythonista** [comment]
@dsaxton it happens often to many people, but I'm not quite sure what's the cause.

**Comment by dsaxton** [comment]
@jreback NP, happy to help

**Pull request merged**